blobs.go

```diff
@@ -0,0 +1,149 @@
+	ErrBlobUploadUnknown = fmt.Errorf("blob upload unknown")
+
+	// ErrBlobInvalidLength returned when the blob has an expected length on
+	// commit, meaning mismatched with the descriptor or an invalid value.
+	ErrBlobInvalidLength = fmt.Errorf("blob invalid length")
+)
+
+// ErrBlobInvalidDigest returned when digest check fails.
+type ErrBlobInvalidDigest struct {
+	Digest digest.Digest
+	Reason error
+}
+
+func (err ErrBlobInvalidDigest) Error() string {
+	return fmt.Sprintf("invalid digest for referenced layer: %v, %v",
+		err.Digest, err.Reason)
+}
+
+// Descriptor describes targeted content. Used in conjunction with a blob
+// store, a descriptor can be used to fetch, store and target any kind of
+// blob. The struct also describes the wire protocol format. Fields should
+// only be added but never changed.
+type Descriptor struct {
+	// MediaType describe the type of the content. All text based formats are
+	// encoded as utf-8.
+	MediaType string `json:"mediaType,omitempty"`
+
+	// Length in bytes of content.
+	Length int64 `json:"length,omitempty"`
+
+	// Digest uniquely identifies the content. A byte stream can be verified
+	// against against this digest.
+	Digest digest.Digest `json:"digest,omitempty"`
+
+	// NOTE: Before adding a field here, please ensure that all
+	// other options have been exhausted. Much of the type relationships
+	// depend on the simplicity of this type.
+}
+
+// BlobDescriptorService makes blob descriptor data available by digest. The
+// service may provide a descriptor of a different digest if the provided
+// digest is not canonical.
+type BlobDescriptorService interface {
+	// Stat provides metadata about a blob identified by the digest.
+	Stat(ctx context.Context, dgst digest.Digest) (Descriptor, error)
+}
+
+// ReadSeekCloser is the primary reader type for blob data, combining
+// io.ReadSeeker with io.Closer.
+type ReadSeekCloser interface {
+	io.ReadSeeker
+	io.Closer
+}
+
+// BlobReadService describes operations for getting blob data.
+type BlobReadService interface {
+	// Get returns the entire blob identified by digest along with the descriptor.
+	Get(ctx context.Context, dgst digest.Digest) ([]byte, error)
+
+	// Open provides a ReadSeekCloser to the blob identified by the provided
+	// descriptor. If the blob is not known to the service, an error will be
+	// returned.
+	Open(ctx context.Context, desc Descriptor) (ReadSeekCloser, error)
+
+	// ServeBlob attempts to serve the blob, identifed by desc, via http. The
+	// service may decide to redirect the client elsewhere or serve the data
+	// directly. This handler only deals with data or redirect. If the blob is
+	// not available, an error will be returned and the caller must still
+	// serve a response. The implementation may serve the same blob from a
+	// different digest domain but the provided media type, if provided, will
+	// be honored.
+	ServeBlob(ctx context.Context, w http.ResponseWriter, r *http.Request, desc Descriptor) error
```

Comment:
Consider not using `Descriptor` here, rather than digest. We do get an optimization here by passing in the length and mediatype, but at the same time, this could be set by the caller.